Makefile

```
@@ -1,37 +1,75 @@
+include .bingo/Variables.mk
 SHELL=/bin/bash
 
-GODIRS_NOVENDOR   ?= $(shell go list ./... | grep -v /vendor/)
+PROVIDER_MODULES ?= $(shell ls -d $(PWD)/providers/*)
+MODULES          ?= $(PROVIDER_MODULES) $(PWD)/
+
 GOFILES_NOVENDOR  ?= $(shell find . -path ./vendor -prune -o -name '*.go' -print)
 
-GOPATH            ?= $(shell go env GOPATH)
 GOBIN             ?= $(firstword $(subst :, ,${GOPATH}))/bin
+
 GO111MODULE       ?= on
 export GO111MODULE
 GOPROXY           ?= https://proxy.golang.org
 export GOPROXY
 
-# TODO(bwplotka): Auto install using separate go.mod for tools.
-GOIMPORTS         ?= $(GOBIN)/goimports
+.PHONY: fmt
+fmt:
+	@echo "Running fmt for all modules: ${MODULES}"
+	@$(GOIMPORTS) -local github.com/grpc-ecosystem/go-grpc-middleware/v2 -w $(GOFILES_NOVENDOR)
+	@$(foreach dir,$(MODULES),$(MAKE) fmt_module DIR=$(dir))
 
-all: vet fmt test
+.PHONY: fmt_module
+fmt_module:
+	go fmt $(shell cd $(DIR) && go list ./... | grep -v /vendor/)
 
-fmt:
-	go fmt $(GODIRS_NOVENDOR)
-	@$(GOIMPORTS) -local github.com/grpc-ecosystem/go-grpc-middleware/v2 -w ${GOFILES_NOVENDOR}
-
-vet:
-	# Do not check lostcancel, they are intentional.
-	# TODO(bwplotka): Explain why intentional.
-	go vet -lostcancel=false $(GODIRS_NOVENDOR)
-
-proto: ./grpctesting/testpb/test.proto
-	@scripts/genproto.sh
-
-# TODO(bwplotka): This depends on test_proto, but CI does not have it, so let's skip it for now.
-test: vet
-	cd ./providers/kit/ && go test -v -race ./...
-	cd ./providers/logrus/ && go test -v -race ./...
-	cd ./providers/zap/ && go test -v -race ./...
+.PHONY: proto
+proto: $(PROTOC_GEN_GOGOFAST) $(GOIMPORTS) $(PROTOC) ./grpctesting/testpb/test.proto
+	@GOIMPORTS_BIN="$(GOIMPORTS)" PROTOC_BIN="$(PROTOC)" scripts/genproto.sh
+
+.PHONY: test
+test:
+	@echo "Running tests for all modules: ${MODULES}"
+	@$(foreach dir,$(PROVIDER_MODULES),$(MAKE) test_module DIR=$(dir))
 	./scripts/test_all.sh
 
-.PHONY: all test
+.PHONY: test_module
+test_module:
+	cd $(DIR) && go test -v -race ./...
+
+.PHONY: lint
+lint:
+	@echo "Running lint for all modules: ${MODULES}"
+	@$(foreach dir,$(MODULES),$(MAKE) test_module DIR=$(dir))
+
+.PHONY: lint_module
+# PROTIP:
+# Add
+#      --cpu-profile-path string   Path to CPU profile output file
+#      --mem-profile-path string   Path to memory profile output file
+# to debug big allocations during linting.
+lint: ## Runs various static analysis against our code.
+lint_module: $(FAILLINT) $(GOLANGCI_LINT) $(MISSPELL) fmt_module
+	$(call require_clean_work_tree,"detected not clean master before running lint")
+	@echo ">> verifying modules being imported"
+	@$(FAILLINT) -paths "errors=github.com/pkg/errors" ./...
+	@$(FAILLINT) -paths "fmt.{Print,PrintfPrintln,Sprint}" -ignore-tests ./...
```

Can we merge these runs

---

Do you want to block fmts in tests? then yes (:

---

I suppose we can use `t.Logf` in tests? Also, should it be
```suggestion
	@$(FAILLINT) -paths "errors=github.com/pkg/errors" ./...
	@$(FAILLINT) -paths "fmt.{Print,Printf,Println,Sprint}" -ignore-tests ./...
```

---

Can agree (: